### DISCUSSION TITLE
Filter for wired Interfaces?

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#17973] [Q&A]
How can I filter for wired (non-wireless) interfaces?  The `is_wired` property does not seem to be available as a filter.  The only way I thought of is to use `Q` with a list of all wired types, which is very unwieldy:

```python
from django.db.models import Q

Interface.objects.filter(
    Q(type='100base-fx')|
    Q(type='100base-lfx')|
    Q(type='100base-tx')|
    Q(type='100base-t1')|
    Q(type='1000base-tx')|
    Q(type='2.5gbase-t')|
    Q(type='5gbase-t')|
    Q(type='10gbase-t')|
    Q(type='10gbase-cx4')|
    Q(type='100base-x-sfp')|
    Q(type='1000base-x-gbic')|
    Q(type='1000base-x-sfp')|
    Q(type='10gbase-x-sfpp')|
    Q(type='10gbase-x-xfp')|
    Q(type='10gbase-x-xenpak')|
    Q(type='10gbase-x-x2')|
    Q(type='25gbase-x-sfp28')|
    Q(type='50gbase-x-sfp56')|
    Q(type='40gbase-x-qsfpp')|
    Q(type='50gbase-x-sfp28')|
    Q(type='100gbase-x-cfp')|
    Q(type='100gbase-x-cfp2')|
    Q(type='200gbase-x-cfp2')|
    Q(type='400gbase-x-cfp2')|
    Q(type='100gbase-x-cfp4')|
    Q(type='100gbase-x-cxp')|
    Q(type='100gbase-x-cpak')|
    Q(type='100gbase-x-dsfp')|
    Q(type='100gbase-x-sfpdd')|
    Q(type='100gbase-x-qsfp28')|
    Q(type='100gbase-x-qsfpdd')|
    Q(type='200gbase-x-qsfp56')|
    Q(type='200gbase-x-qsfpdd')|
    Q(type='400gbase-x-qsfp112')|
    Q(type='400gbase-x-qsfpdd')|
    Q(type='400gbase-x-osfp')|
    Q(type='400gbase-x-osfp-rhs')|
    Q(type='400gbase-x-cdfp')|
    Q(type='400gbase-x-cfp8')|
    Q(type='800gbase-x-qsfpdd')|
    Q(type='800gbase-x-osfp')
)
```

**Comment by candlerb** [comment]
Firstly, you can filter on a list, which simplifies the query a lot:

```
Interface.objects.filter(type__in=['1000base-t','2.5gbase-t'])
```

It's better to use the constants, because then you can't accidentally pick an invalid one:

```
Interface.objects.filter(type__in=[InterfaceTypeChoices.TYPE_1GE_FIXED, InterfaceTypeChoices.TYPE_2GE_FIXED])
```

Then you can use the pre-defined [choice sets](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/dcim/choices.py#L1019-L1224) to get the whole set of interfaces of a particular type:

```
ETHERNETS = [c for (label, choices) in InterfaceTypeChoices.CHOICES if "Ethernet" in label for (c, _) in choices]
Interface.objects.filter(type__in=ETHERNETS)
```

That doesn't give you all "wired" interfaces (it won't include FibreChannel, Infiniband, ATM etc). If you really want all non-wireless:

 ```
NON_WIRELESS = [c for (label, choices) in InterfaceTypeChoices.CHOICES if label != "Wireless" for (c, _) in choices]
Interface.objects.filter(type__in=NON_WIRELESS)
```

Of course, you're hoping that Netbox will never change the human-readable labels on the choice sets, but I think that's as good as you can get. There's an [Interface Kind](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/dcim/choices.py#L848-L857) but I can't see any way that's associated to the interface type.

**Reply by candlerb** [reply]
> The `is_wired` property does not seem to be available as a filter

I forgot to check that. I found it uses additional constants WIRELESS_IFACE_TYPES and VIRTUAL_IFACE_TYPES (and NONCONNECTABLE_IFACE_TYPES which is the union of those). They are in [netbox/dcim/constants.py](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/dcim/constants.py#L38-L58) rather than netbox/dcim/choices.py

This gives another, safer option which exactly matches the logic in `is_wired`:

```
Interface.objects.exclude(type__in=NONCONNECTABLE_IFACE_TYPES)
```
